Python/fileutils.c

```diff
@@ -1804,6 +1812,62 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
             }
         }
     }
+
+    // In non-blocking mode, a write that exceeds the available size of a pipe
+    // fails without writing anything. Limiting writes to the pipe size allows
+    // a buffered write to succeed eventually, as the pipe is read.
+    HANDLE hfile = _Py_get_osfhandle(fd);
+    if (hfile == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+    DWORD pipe_mode, pipe_flags, pipe_size, pipe_insize, pipe_outsize;
+    if (gil_held) {
+        Py_BEGIN_ALLOW_THREADS
+        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
+            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
+                                     NULL, 0) &&
+            pipe_mode & PIPE_NOWAIT)
+        {
+            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
+            // CreatePipe() includes this access for the write handle.
+            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
+                                  &pipe_insize, NULL))
+            {
+                pipe_size = 4096;
+            }
+            else {
+                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
+                                pipe_insize;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+        }
+        Py_END_ALLOW_THREADS
+    }
+    else {
+        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
+            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
+                                     NULL, 0) &&
+            pipe_mode & PIPE_NOWAIT)
+        {
+            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
+            // CreatePipe() includes this access for the write handle.
+            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
+                                  &pipe_insize, NULL))
+            {
+                pipe_size = 4096;
+            }
+            else {
+                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
+                                pipe_insize;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+        }
+    }
```

Comment:
As discussed, retry using a loop and decreasing count instead of capping the write at the pipe size.

```suggestion
```